File: Sources/BambuBar/Services/MoonrakerClient.swift
import Foundation

/// Polls a Klipper printer's Moonraker API and reports status through MQTTClient.Event, so
/// PrinterStore treats Klipper and Bambu connections identically. REST polling keeps the first
/// version simple; a WebSocket subscription can replace the loop later.
final class MoonrakerClient: PrinterConnection, @unchecked Sendable {
    private let printer: SavedPrinter
    private let onEvent: @Sendable (MQTTClient.Event) -> Void
    private var task: Task<Void, Never>?
    private var telemetry = PrinterTelemetry()
    private var connectedReported = false
    private var disconnectReported = false

    init(printer: SavedPrinter, onEvent: @escaping @Sendable (MQTTClient.Event) -> Void) {
        self.printer = printer
        self.onEvent = onEvent
    }

    func start() {
        task = Task.detached(priority: .utility) { [weak self] in await self?.run() }
    }

    func stop() { task?.cancel() }

    private var baseURL: String { "http://\(printer.host):\(printer.port ?? 7125)" }

    private func run() async {
        guard let queryURL = await buildQueryURL() else {
            reportDisconnected("Nie znaleziono API Moonraker (port \(printer.port ?? 7125))")
            return
        }
        while !Task.isCancelled {
            do {
                let data = try await get(queryURL)
                if let updated = MoonrakerStatusParser.telemetry(from: data, previous: telemetry) {
                    telemetry = updated
                    if !connectedReported { connectedReported = true; onEvent(.connected) }
                    onEvent(.telemetry(updated))
                }
            } catch is CancellationError {
                return
            } catch {
                reportDisconnected(error.localizedDescription)
                return
            }
            try? await Task.sleep(for: .seconds(2))
        }
    }

    /// Query only objects the printer actually exposes, so Moonraker doesn't reject the request,
    /// and pick up any chamber temperature sensor and the MMU object if present.
    private func buildQueryURL() async -> URL? {
        var wanted = ["print_stats", "virtual_sdcard", "display_status", "extruder", "heater_bed", "mmu"]
        if let listURL = URL(string: "\(baseURL)/printer/objects/list"),
           let data = try? await get(listURL),
           let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
           let objects = (root["result"] as? [String: Any])?["objects"] as? [String] {
            let available = Set(objects)
            wanted = wanted.filter { available.contains($0) }
            for object in objects where object.lowercased().contains("chamber")
                && (object.hasPrefix("temperature_sensor") || object.hasPrefix("heater_generic")) {
                wanted.append(object)
            }
        }
        guard !wanted.isEmpty else { return nil }
        var components = URLComponents(string: "\(baseURL)/printer/objects/query")
        components?.queryItems = wanted.map { URLQueryItem(name: $0, value: nil) }
        return components?.url
    }

    private func get(_ url: URL) async throws -> Data {
        var request = URLRequest(url: url)
        request.timeoutInterval = 8
        request.cachePolicy = .reloadIgnoringLocalCacheData
        if let key = printer.apiKey, !key.isEmpty { request.setValue(key, forHTTPHeaderField: "X-Api-Key") }
        let (data, response) = try await URLSession.shared.data(for: request)
        guard (response as? HTTPURLResponse)?.statusCode == 200 else { throw URLError(.badServerResponse) }
        return data
    }

    private func reportDisconnected(_ reason: String?) {
        guard !disconnectReported else { return }
        disconnectReported = true
        onEvent(.disconnected(reason))
    }
}

File: windows/BambuBar.Windows/Services/MoonrakerClient.cs
using System.Net.Http;
using System.Text.Json;
using BambuBar.Models;

namespace BambuBar.Services;

/// <summary>Polls a Klipper printer's Moonraker API and reports status through MqttEvent, so
/// PrinterStore treats Klipper and Bambu connections identically. Mirrors the macOS client.</summary>
public sealed class MoonrakerClient : IPrinterConnection
{
    private static readonly HttpClient Http = new() { Timeout = TimeSpan.FromSeconds(8) };

    private readonly SavedPrinter _printer;
    private readonly Action<MqttEvent> _onEvent;
    private readonly CancellationTokenSource _cts = new();
    private PrinterTelemetry _telemetry = new();
    private bool _connectedReported;
    private bool _disconnectReported;

    public MoonrakerClient(SavedPrinter printer, Action<MqttEvent> onEvent)
    {
        _printer = printer;
        _onEvent = onEvent;
    }

    public void Start() => _ = Task.Run(RunAsync);

    public void Stop()
    {
        try { _cts.Cancel(); } catch { }
    }

    private string BaseUrl => $"http://{_printer.Host}:{_printer.Port ?? 7125}";

    private async Task RunAsync()
    {
        var query = await BuildQueryUrlAsync();
        if (query is null)
        {
            ReportDisconnected(AppSettings.Text($"Nie znaleziono API Moonraker (port {_printer.Port ?? 7125})",
                $"Moonraker API not found (port {_printer.Port ?? 7125})"));
            return;
        }
        while (!_cts.IsCancellationRequested)
        {
            try
            {
                var data = await GetAsync(query);
                var updated = MoonrakerStatusParser.Telemetry(data, _telemetry);
                if (updated is not null)
                {
                    _telemetry = updated;
                    if (!_connectedReported) { _connectedReported = true; _onEvent(new MqttEvent { Type = MqttEventType.Connected }); }
                    _onEvent(new MqttEvent { Type = MqttEventType.Telemetry, Telemetry = updated });
                }
            }
            catch (OperationCanceledException) { return; }
            catch (Exception ex) { ReportDisconnected(ex.Message); return; }
            try { await Task.Delay(TimeSpan.FromSeconds(2), _cts.Token); }
            catch (OperationCanceledException) { return; }
        }
    }

    private async Task<string?> BuildQueryUrlAsync()
    {
        var wanted = new List<string> { "print_stats", "virtual_sdcard", "display_status", "extruder", "heater_bed", "mmu" };
        try
        {
            var listData = await GetAsync($"{BaseUrl}/printer/objects/list");
            using var doc = JsonDocument.Parse(listData);
            if (doc.RootElement.TryGetProperty("result", out var res)
                && res.TryGetProperty("objects", out var objects) && objects.ValueKind == JsonValueKind.Array)
            {
                var available = objects.EnumerateArray().Select(o => o.GetString() ?? "").ToHashSet();
                wanted = wanted.Where(available.Contains).ToList();
                foreach (var name in available)
                    if (name.ToLowerInvariant().Contains("chamber")
                        && (name.StartsWith("temperature_sensor") || name.StartsWith("heater_generic")))
                        wanted.Add(name);
            }
        }
        catch { /* fall back to the default object set */ }

        if (wanted.Count == 0) return null;
        var query = string.Join("&", wanted.Select(Uri.EscapeDataString));
        return $"{BaseUrl}/printer/objects/query?{query}";
    }

    private async Task<byte[]> GetAsync(string url)
    {
        using var request = new HttpRequestMessage(HttpMethod.Get, url);
        if (!string.IsNullOrEmpty(_printer.ApiKey)) request.Headers.Add("X-Api-Key", _printer.ApiKey);
        using var response = await Http.SendAsync(request, _cts.Token);
        response.EnsureSuccessStatusCode();
        return await response.Content.ReadAsByteArrayAsync(_cts.Token);
    }

    private void ReportDisconnected(string? reason)
    {
        if (_disconnectReported) return;
        _disconnectReported = true;
        _onEvent(new MqttEvent { Type = MqttEventType.Disconnected, Reason = reason });
    }
}
